tevfik_aksoy/graph.cpp

```diff
@@ -0,0 +1,172 @@
+  }
+  if (source.id == destination.id)
+    return Edge::Color::Green;
+  if (source.depth == destination.depth) {
+    for (int i = 0; i < depth_map_[source.depth].size() - 1; i++) {
+      const auto first = depth_map_[source.depth][i];
+      const auto second = depth_map_[source.depth][i + 1];
+      if ((source.id == first && destination.id == second) ||
+          (destination.id == first && source.id == second))
+        return Edge::Color::Blue;
+    }
+  }
+  if (source.depth == destination.depth - 1)
+    return Edge::Color::Yellow;
+  if (source.depth == destination.depth - 2)
+    return Edge::Color::Red;
+
+  throw std::runtime_error("Failed to calculate edge color");
+}
+
+Vertex Graph::get_vertex(const VertexId& id) const {
+  for (const auto& vertex : vertices_) {
+    if (id == vertex.id)
+      return vertex;
+  }
+  throw std::runtime_error("Vertex not found!");
+}
+
+void Graph::insert_edge(const VertexId& source_id,
+                        const VertexId& destination_id) {
+  assert(!are_vertices_connected(source_id, destination_id) &&
+         "Vertices are already connected!");
+  const auto& source_vertex = get_vertex(source_id);
+  const auto& destination_vertex = get_vertex(destination_id);
+  const auto color =
+      calculate_color_for_edge(source_vertex, destination_vertex);
+  const int edge_id = get_new_edge_id();
+  colored_edges_map_[color].push_back(edge_id);
+  edges_.emplace_back(source_id, destination_id, edge_id, color);
+
+  vertices_[source_id].add_edge_id(edge_id);
+  if (color != Edge::Color::Green) {
+    vertices_[destination_id].add_edge_id(edge_id);
+    if (color == Edge::Color::Gray) {
+      const auto depth = vertices_[source_id].depth + 1;
+      vertices_[destination_id].depth = depth;
+      if (depth_map_.size() == depth) {
+        depth_map_.emplace_back();
+      }
+      depth_map_[depth].emplace_back(destination_id);
+    }
+  }
+}
+
+bool Graph::are_vertices_connected(const VertexId& source,
+                                   const VertexId& destination) const {
+  assert(does_vertex_exist(source, vertices_) &&
+         "Source vertex doesn't exist!");
+  assert(does_vertex_exist(destination, vertices_) &&
+         "Destination vertex doesn't exist!");
+
+  const auto& source_vertex_edges = vertices_[source].get_edge_ids();
+  const auto& destination_vertex_edges = vertices_[destination].get_edge_ids();
+  for (const auto& edge_of_source_vertex : source_vertex_edges)
+    if (source == destination) {
+      if (edges_[edge_of_source_vertex].source ==
+          edges_[edge_of_source_vertex].destination)
+        return true;
+    } else {
+      for (const auto& edge_of_destination_vertex : destination_vertex_edges)
+        if (edge_of_source_vertex == edge_of_destination_vertex)
+          return true;
+    }
+  return false;
+}
+
+const std::vector<EdgeId>& Graph::get_colored_edges(
+    const Edge::Color& color) const {
+  return colored_edges_map_.at(color);

```

Comment:
Да, ты должен быть уверен что тут не будет `exception`. Так как мы всегда можем запросить цвет, и в случае, если граней нет, мы хотим просто вернуть пустой вектор - вот один из вариантов,Ю как этого добиться:
```suggestion
  if (colored_edges_map_.find(color) == colored_edges_map_.end()) {
    static std::vector<EdgeId> empty_result;
    return empty_result;
  }
  return colored_edges_map_.at(color);
```

Reply:
Понял, спасибо!